python/dgl/transform.py

```diff
@@ -208,22 +222,102 @@ def segmented_knn_graph(x, k, segs):
     (tensor([0, 0, 1, 1, 1, 2, 3, 3, 4, 4, 5, 5, 6, 6]),
      tensor([0, 1, 0, 1, 2, 2, 3, 5, 4, 6, 3, 5, 4, 6]))
     """
-    n_total_points, _ = F.shape(x)
-    offset = np.insert(np.cumsum(segs), 0, 0)
-
-    h_list = F.split(x, segs, 0)
-    dst = [
-        F.argtopk(pairwise_squared_distance(h_g), k, 1, descending=False) +
-        int(offset[i])
-        for i, h_g in enumerate(h_list)]
-    dst = F.cat(dst, 0)
-    src = F.arange(0, n_total_points).unsqueeze(1).expand(n_total_points, k)
-
-    dst = F.reshape(dst, (-1,))
-    src = F.reshape(src, (-1,))
-    adj = sparse.csr_matrix((F.asnumpy(F.zeros_like(dst) + 1), (F.asnumpy(dst), F.asnumpy(src))))
-
-    return convert.from_scipy(adj)
+    if use_dgl_impl:
+        out = knn(x, segs, x, segs, k).t().contiguous()
+        row, col = out[1], out[0]
+        return convert.graph((row, col))
+    else:
+        n_total_points, _ = F.shape(x)
+        offset = np.insert(np.cumsum(segs), 0, 0)
+
+        h_list = F.split(x, segs, 0)
+        dst = [
+            F.argtopk(pairwise_squared_distance(h_g), k, 1, descending=False) +
+            int(offset[i])
+            for i, h_g in enumerate(h_list)]
+        dst = F.cat(dst, 0)
+        src = F.arange(0, n_total_points).unsqueeze(1).expand(n_total_points, k)
+
+        dst = F.reshape(dst, (-1,))
+        src = F.reshape(src, (-1,))
+        adj = sparse.csr_matrix((F.asnumpy(F.zeros_like(dst) + 1), (F.asnumpy(dst), F.asnumpy(src))))
+
+        return convert.from_scipy(adj)
+
+def knn(x, x_segs, y, y_segs, k, dist:str="euclidean"):
+    r"""For each element in each segment in :attr:`y`, find :attr:`k` nearest
+    points in the same segment in :attr:`x`.
+
+    This function allows multiple point sets with different capacity. The points
+    from different sets are stored contiguously in the :attr:`x` and :attr:`y` tensor.
+    :attr:`x_segs` and :attr:`y_segs` specifies the number of points in each point set.
+
+    Parameters
+    ----------
+    x : Tensor
+        The point coordinates in x. It can be either on CPU or GPU (must be the
```

Comment:
Do we support GPU now? If not then we shouldn't claim like this in the doc. Also, we need to add corresponding checks below.

Reply:
Currently, if the input tensor is not on the CPU, we will perform a copy operation.

Reply:
This behavior should definitely let user know. Either we ask users to copy to CPU themselves, or at least print a warning as the notification. GPU to CPU copy is slow, and I don't see the code copying the result back to GPU -- this will also break user experience.

Reply:
`knn_graph` and `segmented_knn_graph` return cpu graphs, regardless of the context of inputs. I also add a warning in `knn` if input tensors are not on cpu

Reply:
So the input features are copied to CPU before brute-force?  If so we can just do the same and always use KD-tree once we prove that KD-tree is better than brute-force.

Reply:
If use GPU and brute-force, kNN results will be first computed, then the results will be copied to CPU for graph construction.